dvc/command/experiments.py

```diff
@@ -471,9 +473,21 @@ def show_experiments(
         }
     )
 
-    if kwargs.get("only_changed", False):
+    if kwargs.get("only_changed", False) or html:
         td.drop_duplicates("cols")
 
+    extra_render_args = {}
+    if html:
+        td.dropna("rows")
```

Comment:
> > Anyhow would you expect to see `foo` in the plot if there is only 1 value?
> 
> That might be okay for now, although eventually we will need a better solution since it means missing a field for a single experiment makes it invisible, right? We might need an N/A value for each field where some data is missing, or we can just have a missing point and corresponding gap in the line for experiments that have no value for that field.

I fixed 2 separated bugs:

[One](https://github.com/iterative/dvc/pull/7064) was unexpectedly making `foo` slip through the `only_changed` filter. So, in the example you shared before, the `foo` column won't be shown on the plot (whether this is desirable or not doesn't affect the fix).

The other was related to how `NaNs` were treated in categorical columns (unique values were being assigned to each NaN). The fix is now included as part of this p.r. making NaNs converge to a single point.

If we didn't filter the `foo` column per the first fix, it would look like the following:

```dvc exp show -T --html --sort-by foo```

![newplot(3)](https://user-images.githubusercontent.com/12677733/143948236-3492ad51-e04c-498c-aa5d-9c73af6a5b70.png)

Reply:
> That might be okay for now, although eventually we will need a better solution since it means missing a field for a single experiment makes it invisible, right?

There are 2 different filters:

First, ` td.drop_duplicates("cols")` removes any columns with 0 variance. Here we could discuss if we want to consider `NaNs` as unique values (keep `foo` in the example`) or not, I don't have a strong opinion here.

After that, `td.dropna("rows")` drops rows with *any* NaNs. This was my decision to prevent populating the plot with useless lines (i.e. commits missing some metric/param) but also because NaNs are very problematic for scalar columns as it totally messes up the plot (scalar columns are converted to categorical if NaN is present). NaNs in categorical columns are ok.

We could rework the second filter in many ways:

- Using `all` option instead of `any` (works for useless lines but does not solve scalar columns with NaNs).
- Adding a new `thresh` option mimicking pandas (magic number hard to choose ).
- Being smart about choosing the fill value for NaNs in scalar columns.

> or we can just have a missing point and corresponding gap in the line for experiments that have no value for that field.

This is not supported by plotly.

Reply:
@dberenbaum Here is the latest state:

- Rows are only dropped if *all* of the columns are NaN.

- `NaN` in **categorical** columns are  represented with a `Missing` value. This makes NaN lines converge to a single point:

![newplot(13)](https://user-images.githubusercontent.com/12677733/144317387-7d572c47-f833-4e9e-b536-cbfa30a3454a.png)

- `NaN` in **scalar** columns (creating `foo: 1` in the example repo) are represented as `null`, leaving the behavior dependent on `plotly` internals (for better or for worse). This makes NaN lines converge to a point beyond the limits of the plot:

![newplot(12)](https://user-images.githubusercontent.com/12677733/144317447-d2775a11-dd65-4eaf-b153-274fe27fa74c.png)

Reply:
Opened https://github.com/iterative/dvc/issues/7083 for improving the `Rows are only dropped if all of the columns are NaN` to only consider metrics and params columns.